java/runtime/src/main/java/io/ray/runtime/object/ObjectRefImpl.java

```diff
@@ -53,4 +51,53 @@ public ObjectId getId() {
     return type;
   }
 
+  @Override
+  protected void finalize() throws Throwable {
+    try {
+      removeLocalReference();
+    } finally {
+      super.finalize();
+    }
+  }
+
+  static ThreadLocal<Set<ObjectId>> containedObjectIds = ThreadLocal.withInitial(HashSet::new);
+
+  public static List<ObjectId> getAndClearContainedObjectIds() {
+    List<ObjectId> ids = new ArrayList<>(containedObjectIds.get());
+    containedObjectIds.get().clear();
+    return ids;
+  }
+
+  @Override
+  public void writeExternal(ObjectOutput out) throws IOException {
+    out.writeObject(this.getId());
+    out.writeObject(this.getType());
+    containedObjectIds.get().add(this.getId());
+  }
+
+  @Override
+  public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+    this.id = (ObjectId) in.readObject();
+    this.type = (Class<T>) in.readObject();
+    addLocalReference();
+  }
+
+  private void addLocalReference() {
+    Preconditions.checkState(workerId == null);
+    RayRuntimeInternal runtime = (RayRuntimeInternal) Ray.internal();
+    workerId = runtime.getWorkerContext().getCurrentWorkerId();
+    runtime.getObjectStore().addLocalReference(workerId, id);
+  }
+
+  // This method is public for test purposes only.
+  public synchronized void removeLocalReference() {
```

Comment:
I know the comment says it's only for test purposes, but it'd be great to not have this method be public. If we can make all the test cases work with JVM GC only, that would be best.

Reply:
Thanks for the advice. I'll make this method private and let UT invoking it by reflection.